crates/rattler_installs_packages/src/index/package_sources.rs

```diff
@@ -0,0 +1,108 @@
+            base_source: base_index_url,
+            extra_sources: Default::default(),
+            overrides: Default::default(),
+        }
+    }
+
+    /// Add another index URL
+    pub fn with_index(&mut self, alias: String, url: Url) -> &mut Self {
+        self.extra_sources.push(PackageSource { alias, url });
+        self
+    }
+
+    /// Add an override for a specific package. This will cause the package to be installed
+    /// from the given source and from that source only
+    pub fn with_override(&mut self, package: NormalizedPackageName, alias: String) -> &mut Self {
+        self.overrides.insert(package, alias);
+        self
+    }
+
+    /// Finalize the builder and create a `PackageSources` instance
+    pub fn build(&self) -> miette::Result<PackageSources> {
+        let mut extra_sources_map = BTreeMap::new();
+        self.extra_sources
+            .iter()
+            .enumerate()
+            .map(|(i, source)| {
+                // Offset by one because index_urls will include the default source
+                (source.alias.clone(), i + 1)
+            })
+            .try_for_each(|(alias, index)| {
+                if extra_sources_map.insert(alias.clone(), index).is_some() {
+                    miette::bail!("duplicate index alias '{}'", alias);
+                }
+                Ok(())
+            })?;
+
+        let mut artifact_to_index = BTreeMap::new();
+        self.overrides.iter().try_for_each(|(package, source)| {
+            let index = *extra_sources_map
+                .get(source)
+                .ok_or_else(|| miette::miette!("unknown index alias '{}'", source))?;
+
+            if artifact_to_index.insert(package.clone(), index).is_some() {
+                miette::bail!("duplicate package-source map entry '{}'", package);
+            }
+
+            Ok(())
+        })?;
+
+        let index_urls = std::iter::once(self.base_source.clone())
+            .chain(self.extra_sources.iter().map(|source| source.url.clone()))
+            .collect::<Vec<_>>();
+
+        Ok(PackageSources {
+            index_urls,
+            artifact_to_index,
+        })
+    }
+}
+
+/// A collection of package sources and source overrides.
+/// See [`PackageSourcesBuilder`] for creating an instance of this type.
+pub struct PackageSources {
+    index_urls: Vec<Url>,
+    artifact_to_index: BTreeMap<NormalizedPackageName, usize>,
+}
+
+impl PackageSources {
+    /// Get the index URL for a package
+    pub fn index_url(&self, package: &NormalizedPackageName) -> Vec<&Url> {
+        let maybe_index = self
+            .artifact_to_index
+            .get(package)
+            .map(|&index| &self.index_urls[index]);
+
+        if let Some(url) = maybe_index {
+            vec![url]
+        } else {
+            self.index_urls.iter().collect()
+        }
+    }
+
+    /// Get the default (fallback) index URL
+    pub fn default_index_url(&self) -> Url {
+        self.index_urls.first().unwrap().clone()
```

Comment:
Please never use unwrap (except for tests) but prefer `expect` with an explanation why this will never be hit. That provides readers and future debuggers with your assumptions.